src/CSnakes.SourceGeneration/Parser/EnumerableExtensions.cs

```diff
@@ -31,4 +31,37 @@ internal static partial class EnumerableExtensions
             }
         }
     }
+
+    public static IEnumerable<int[]> Permutations(this IEnumerable<int> ranges)
+    {
+        // Given a list of ranges, e.g.
+        // [2, 3, 1] (2 options for first, 3 for second, 1 for third), return all combinations of indexes
+        // e.g. [0, 0, 0], [0, 0, 1], [0, 1, 0], [0, 1, 1], [1, 0, 0], [1, 0, 1], [1, 1, 0], [1, 1, 1]
```

Comment:
Shouldn't this read as follows?

```suggestion
        // e.g. [0, 0, 0] [0, 1, 0] [0, 2, 0] [1, 0, 0] [1, 1, 0] [1, 2, 0]
```